### ISSUE TITLE
TF reports cublas64_10.dll not found

### ISSUE THREAD

**Issue by SamFigueroa** [tensorflow/tensorflow#40108]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 Version 1909
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.2
- Python version: 3.8.3
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1/7.6.5
- GPU model and memory: NVIDIA Quadro M1200 - 4 GB

**Describe the problem**
I am trying to get tensorflow to make use of the NVIDIA GPU in my laptop.  Tensorflow is able to find all CUDA-related libraries *except* cublas64_10.dll.  After installing CUDA 10.1 (I've tried 10.1 and 10.1 update 2, both without any success), I copied all the .dll files that got installed in "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin" to C:\Windows\System32.  I included cublas64_10.dll when I copied these files.  (Before copying the .dll files, tensorflow could not find any of the CUDA-related libraries, even after adding "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin" to my PATH and rebooting my computer a couple times.)  I confirmed that the cublas64_10.dll in C:\Windows\System32 was bit-for-bit identical to the one in "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin", and that I had permission to read cublas64_10.dll.  (I copied this file from C:\Windows\System32 to a different directory to confirm this.)  I also downloaded just CUBLAS from CUDA 10.2, being very careful no other components got installed on my system, and then copied the cublas64_10.dll file from that version of CUDA to C:\Windows\System32.  None of my efforts were fruitful.

**Provide the exact sequence of commands / steps that you executed before running into the problem**
python -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
>python -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"
2020-06-02 21:13:03.752642: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-06-02 21:13:06.095237: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2020-06-02 21:13:06.576837: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro M1200 computeCapability: 5.0
coreClock: 1.148GHz coreCount: 5 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 74.65GiB/s
2020-06-02 21:13:06.588734: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-06-02 21:13:06.594961: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cublas64_10.dll'; dlerror: cublas64_10.dll not found
2020-06-02 21:13:06.604653: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-06-02 21:13:06.615330: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-06-02 21:13:06.626644: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-06-02 21:13:06.635321: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-06-02 21:13:06.652175: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-06-02 21:13:06.658845: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
[]

>DIR "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin\cublas64_10.dll" C:\Windows\System32\cublas64_10.dll
 Volume in drive C is OSDisk
 Volume Serial Number is E295-9E24

 Directory of C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin

07/28/2019  10:30 PM        58,830,848 cublas64_10.dll
               1 File(s)     58,830,848 bytes

 Directory of C:\Windows\System32

07/28/2019  10:30 PM        58,830,848 cublas64_10.dll
               1 File(s)     58,830,848 bytes
               0 Dir(s)  64,581,230,592 bytes free

>PATH
PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\libnvvp;;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\WINDOWS\System32\OpenSSH;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files (x86)\WebEx\Productivity Tools;C:\Program Files\osquery;C:\Program Files\Intel\WiFi\bin;C:\Program Files\Common Files\Intel\WirelessCommon;C:\Program Files\nodejs;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Perforce;C:\Program Files\NVIDIA Corporation\Nsight Compute 2019.4.0\;C:\Users\figueroa\AppData\Local\Microsoft\WindowsApps;C:\Users\figueroa\AppData\Roaming\npm;C:\Users\figueroa\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\Scripts;"C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64";"U:\Perforce\sw\main\apps\p4review";"U:\Perforce\sw\tools\cygnus\bin";"U:\Perforce\sw\tools\win64\ActivePerl\5.10.1.1006\bin"

**Comment by jvishnuvardhan** [comment]
@SamFigueroa Sometimes it feels difficult when compatibility issues arise. I can understand your position. Can you please uninstall completely CUDA as well as TF and then install CUDA and TF2.2. It worked for me when I had similar issue. Thanks!

**Comment by SamFigueroa** [comment]
Would it help if I were to downgrade to TF2.1?

**Comment by jvishnuvardhan** [comment]
@SamFigueroa Requirements are almost same for 2.1 except few as shown below. check [here](https://www.tensorflow.org/install/source_windows) for other versions.


Version | Python version | Compiler | Build tools | cuDNN | CUDA
-- | -- | -- | -- | -- | --
tensorflow_gpu-2.2.0 | 3.5-3.8 | MSVC 2019 | Bazel 2.0.0 | 7.4 | 10.1
tensorflow_gpu-2.1.0 | 3.5-3.7 | MSVC 2019 | Bazel 0.27.1-0.29.1 | 7.4 | 10.1

I would suggest installing `TF2.2` would be better as there were lot of performance improvements in `TF2.2`. Thanks!

**Comment by SamFigueroa** [comment]
Could the version of cuDNN be the problem?  I installed version 7.6.5, which is the latest on NVIDIA's website.  Version 7.4 does not seem to be available for CUDA 10.1.  The most recent version of CUDA for which cuDNN version 7.4.x is available is CUDA 10.0, but that's not the right version for tf 2.1 or 2.2.  Maybe I should use tf 2.0 so I can get all the right versions of CUDA and related libraries that are supposed to work together?

**Comment by SamFigueroa** [comment]
I completely uninstalled tf and CUDA 10.1, and then reinstalled both, but this did not help.  On a hunch that the key clue was that I had to copy .dll files to C:\Windows\System32 in order for tf to find any of them at all, I decided to try the version of Python that came with MS Visual Studio, which is 3.7.4.  I reinstalled tf with this version of Python.  Now I get:

C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin>PYTHON -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"
2020-06-04 21:08:33.482516: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-06-04 21:08:35.007585: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2020-06-04 21:08:35.477814: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro M1200 computeCapability: 5.0
coreClock: 1.148GHz coreCount: 5 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 74.65GiB/s
2020-06-04 21:08:35.487926: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-06-04 21:08:35.498249: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-06-04 21:08:36.138902: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-06-04 21:08:36.406474: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-06-04 21:08:36.452724: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-06-04 21:08:36.931253: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-06-04 21:08:37.000022: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-06-04 21:08:37.006212: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1703] Adding visible gpu devices: 0
[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

I get the same thing, even after deleting all the .dll files I had copied to C:\Windows\System32.  My conclusion, then, is that I need to invest in a better version of Python, as the one I got from the MS Store is worth what I paid for, i.e., -0-.

My next experiment will be to update to CUDA 10.1 Update 2 (I installed plain 10.1 just to be safe), as well as to cuDNN 7.6.5.  (I installed cuDNN 7.4.2, again, just to be safe, but according to NVIDIA's website, this version of cuDNN is not for CUDA 10.1.)  Hopefully updating won't break anything.

**Comment by jvishnuvardhan** [comment]
@SamFigueroa Did you check `CUDA 10.1 Update 2`? From the log above, i see some are from CUDA10 (cublas64_10.dll, curand64_10.dll etc) and one is from CUDA10.1 (cudart64_101.dll). Not sure whether it was completely removed or not. Thanks!

**Comment by Dave-Stein** [comment]
@jgasperlin  
I have been [asking something like this **since 6 months ago**](https://github.com/tensorflow/tensorflow/issues/35405). No progress for anything related to tensorflow2.*

https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-gpu-windows-x86_64-2.0.0.zip


Any chance if you can provide a link to your gpu version of libtensorflow?  So we can continue to test while Google finally makes up the mind to support Windows developers

**Comment by SamFigueroa** [comment]
I updated to CUDA 10.1 Update 2 without uninstalling CUDA, and I also updated to cuDNN 7.6.5, which essentially involved just overwriting the .dll file from the previous version of cuDNN.  I verified that tf was still able to make use of my GPU.  (One way to confirm this beyond just the one-line Python program is to run the training for a model we are working on.  It runs 3-4X faster with the GPU I have, as compared to running just on the CPU.)

Next, I downloaded Python 3.8.3 directly from python.org, and verified all of the above.  I also downloaded TensorRT, but I haven't tried to determine if the GPU is actually being used for inference.  What I can say is that when I run inference, it is finding the GPU in my computer.

In summary, at this point, my Windows 10 laptop is now fully set up for using tensorflow, and it is definitely using the GPU for training, and maybe even for inference.  I look forward to the next version of tensorflow, as that should allow me to use some other systems, such as my Linux desktop, which has CUDA 10.2.  I am unwilling to downgrade to CUDA 10.1 on that system, and my efforts to build tensorflow from source were unsuccessful, but that's a separate story.

**Comment by jvishnuvardhan** [comment]
@SamFigueroa Is it fair to say the issue was resolved? If yes, then please close the issue. For installing TF from source, you could open a new issue so that it is easy for the users to follow the thread. Thanks!

**Comment by SamFigueroa** [comment]
Closing, as this issue is resolved.

**Comment by SamFigueroa** [comment]
Just to summarize the root cause, the Python I got from the Microsoft Store was not able to find the CUDA .dll files I installed.  Using Python from a different source, such as python.org, fixed this issue.

**Comment by HanyuSun** [comment]
I accounted the same problem. After downgrading the version of tensorflow feom 2.3 to 2.2, the problem was solved quickily.

**Comment by a86612** [comment]
Now I also account  the same problem.Follow the author's method to download python3.7 from python.org,but don't solve the problem.

**Comment by a86612** [comment]
I try to download tensorflow2.2.0 version

**Comment by snnn** [comment]
If python 3.7 works but python 3.8 doesn't, please read:

https://github.com/python/cpython/commit/2438cdf0e932a341c7613bf4323d06b91ae9f1f1
https://bugs.python.org/issue36085

Maybe they are relevant.

To verify it, you may use WinDBG and turn on loadsnap debugging , it will show which dir it searched.

**Comment by coffeeshop13** [comment]
Guys, I just found the answer. Somehow in bin folder should be the files: cublasLt64_10.dll and cublas64_10.dll. There will be no error if there is no cublasLt64_10 file there but it must be there. Also if you have cublasLt64_**11** just rename it to cublasLt64_**10**

**Comment by Shadetail** [comment]
> There will be no error if there is no cublasLt64_10 file there but it must be there.
This was exactly the problem I was having and your comment helped me solve it. Thank you so much!

**Comment by barbatoz0220** [comment]
> Guys, I just found the answer. Somehow in bin folder should be the files: cublasLt64_10.dll and cublas64_10.dll. There will be no error if there is no cublasLt64_10 file there but it must be there. Also if you have cublasLt64_**11** just rename it to cublasLt64_**10**

I just had the same issue and your reply helped me alot lol - I'm both happy and pissed at how this can work....
For those of you having the same case where many of the dll files were missing, just download them and add them to your env>Library>bin folder.
Then, download **cublas64_10.dll** online while at the same time renaming cublasLt64_**11** into  cublasLt64_**10** like he/she said.
You shouldn't have any more problems.

**Comment by Ruhail-Ali-Khan** [comment]
> > Guys, I just found the answer. Somehow in bin folder should be the files: cublasLt64_10.dll and cublas64_10.dll. There will be no error if there is no cublasLt64_10 file there but it must be there. Also if you have cublasLt64_**11** just rename it to cublasLt64_**10**
> 
> I just had the same issue and your reply helped me alot lol - I'm both happy and pissed at how this can work....
> For those of you having the same case where many of the dll files were missing, just download them and add them to your env>Library>bin folder.
> Then, download **cublas64_10.dll** online while at the same time renaming cublasLt64_**11** into cublasLt64_**10** like he/she said.
> You shouldn't have any more problems.

I tried it is working but kernel not working when training the model.

**Comment by Drinkers** [comment]
> > Guys, I just found the answer. Somehow in bin folder should be the files: cublasLt64_10.dll and cublas64_10.dll. There will be no error if there is no cublasLt64_10 file there but it must be there. Also if you have cublasLt64_**11** just rename it to cublasLt64_**10**
> 
> I just had the same issue and your reply helped me alot lol - I'm both happy and pissed at how this can work.... For those of you having the same case where many of the dll files were missing, just download them and add them to your env>Library>bin folder. Then, download **cublas64_10.dll** online while at the same time renaming cublasLt64_**11** into cublasLt64_**10** like he/she said. You shouldn't have any more problems.

下载了cublas64_10.dll仍然不行，之后通过下载cublasLt64_10解决了，我的CUDA版本是10.0，tf版本是2.3.0
